### PR TITLE
Hide fields for picture word users

### PR DESCRIPTION
**Pull request by ewjordan**
Trims down account management page for picture/word users to remove e-mail field, disable editing of username, and remove "Delete user" field.

This page should still be inaccessible without knowing the specific URL (/users/edit), a later PR will re-add the button that gives students access to it again.

screen shot 2017-06-08 at 6 14 26 pm

### CODE REVIEW

**Comment by ewjordan** [comment]
@ashercodeorg I could definitely use your feedback on whether this is the right way to check for picture/word accounts, given some of the changes happening there.

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -1252,6 +1252,26 @@ def can_edit_password?
...
+    end
+  end
+
+  def should_show_email?
+    !picture_or_word_account?
+  end
+
+  def can_delete_account?
+    !picture_or_word_account?
+  end
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb]: This adds four methods, where it seems one should suffice. Perhaps `sponsored?` (as suggested above)?

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -1252,6 +1252,26 @@ def can_edit_password?
...
+  def can_edit_display_name?
+    !picture_or_word_account?
+  end
+
+  def picture_or_word_account?
+    return false if sections_as_student.empty?
+    return false if encrypted_password.present?
+    sections_as_student.all? do |section|
+      section.login_type == Section::LOGIN_TYPE_PICTURE || section.login_type == Section::LOGIN_TYPE_WORD
+    end
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb]: My opinion is that this isn't the right way to check for picture and word accounts. In particular, I think `provider == User::PROVIDER_SPONSORED` is both simpler and more correct. Your logic seems to fail in a variety of situations.
* If the account is sponsored and the sponsoring section was deleted.
* If the account is an oauth account belonging to picture sections (only).
* Perhaps others I'm not thinking of off the top of my head.

File: dashboard/app/views/devise/registrations/edit.html.haml [hunk 3]
```diff
@@ -69,12 +73,13 @@
       = f.select :user_type, user_type_options
 
   %div= f.submit t('crud.update'), class: 'btn btn-default'
-%h3= t('user.delete')
-%p
-  = t('user.delete_text')
-  - if current_user.teacher?
-    = t('user.delete_text_teacher_note')
-  = button_to t('user.delete'), registration_path(resource_name), data: { confirm: t('user.delete_confirm') }, method: :delete, class: 'btn btn-danger'
+- if current_user.can_delete_account?
```
Comment (ashercodeorg) [inline dashboard/app/views/devise/registrations/edit.html.haml]: Throughout, the HAML has been using `f.object` rather than `current_user`. Why the change? Should the rest of the HAML be using `current_user`? Should this be using `f.object`?

File: dashboard/app/views/devise/registrations/edit.html.haml [hunk 4]
```diff
@@ -69,12 +73,13 @@
...
+  %h3= t('user.delete')
+  %p
+    = t('user.delete_text')
+    - if current_user.teacher?
+      = t('user.delete_text_teacher_note')
+    = button_to t('user.delete'), registration_path(resource_name), data: { confirm: t('user.delete_confirm') }, method: :delete, class: 'btn btn-danger'
```
Comment (ashercodeorg) [inline dashboard/app/views/devise/registrations/edit.html.haml]: This line is looong. Please shorten.

File: dashboard/test/models/user_test.rb [hunk 5]
```diff
@@ -1036,6 +1036,44 @@ def complete_script_for_user(user, script, completed_date = Time.now)
     assert @student.can_edit_email?
   end
 
+  test 'picture_or_word_account? is false for user in no sections' do
+    refute @student.picture_or_word_account?
+  end
```
Comment (ashercodeorg) [inline dashboard/test/models/user_test.rb]: The statement of the test is wrong.

File: dashboard/test/models/user_test.rb [hunk 6]
```diff
@@ -1036,6 +1036,44 @@ def complete_script_for_user(user, script, completed_date = Time.now)
...
+    student_with_password = create(:student, encrypted_password: '123456')
+    create(:follower, student_user: student_with_password, section: picture_section)
+    student_with_password.reload
+    refute student_with_password.picture_or_word_account?
+
+    # word section
+    student_with_password = create(:student, encrypted_password: '123456')
+    create(:follower, student_user: student_with_password, section: word_section)
+    student_with_password.reload
+    refute student_with_password.picture_or_word_account?
```
Comment (ashercodeorg) [inline dashboard/test/models/user_test.rb]: DRY (also below).
```
[Section::LOGIN_TYPE_PICTURE, Section::LOGIN_TYPE_WORD].each do |login_type|
  section = create :section, login_type: login_type
  follower = create :follower, section: section
  refute follower.student_user.picture_or_word_account?
end
```
The default follower uses the default student, which uses the default user (specifying the user_type), which has a password (so presumably encrypted password).

**Comment by poorvasingal** [comment]
@ryansloan - FYI that this is happening. So you should be able to start adding settings to the accounts page now. 

BTW - to confirm, if a student has an email account but then joins a picture section (meaning they end up with a picture password too), they will not see the restricted set of settings right? In that case, they would have the ability to edit anything a user with an email account can edit?

**Comment by ashercodeorg** [comment]
@poorvasingal, that will be the behavior after this PR, yes.

File: dashboard/app/models/user.rb (hunk 1 above)
Comment (ewjordan) [reply dashboard/app/models/user.rb]: Seems fair - we can always break it out into different conditions if and when that is necessary.

File: dashboard/app/models/user.rb (hunk 2 above)
Comment (ewjordan) [reply dashboard/app/models/user.rb]: OK, I'll dig into this a little bit more, and try to think about what the correct behavior actually should be in those cases, as well.

File: dashboard/app/views/devise/registrations/edit.html.haml (hunk 3 above)
Comment (ewjordan) [reply dashboard/app/views/devise/registrations/edit.html.haml]: Possibly, but I'm a little bit nervous about changing it just to change it - you can't use `f.object` here because it's not in scope for that code, and while it does seem like wherever `f.object` above should be the same as `current_user`, it's coming in as a param to the form block, so technically that would be changing the semantics. While I don't *think* it should make a difference, my (very slight) preference would be to not touch the existing code if it's not broken. Happy to change the `f.object` uses if you feel strongly about it, though.

Comment (ashercodeorg) [reply dashboard/app/views/devise/registrations/edit.html.haml]: That sounds like a legitimate reason, good enough for me. No change required.

File: dashboard/app/models/user.rb [hunk 7]
```diff
@@ -1252,6 +1252,11 @@ def can_edit_password?
     encrypted_password.present?
   end
 
+  def teacher_managed_account?
+    return false if encrypted_password.present?
+    !oauth?
+  end
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb]: Much, much nicer!

**Comment by ewjordan** [comment]
@ashercodeorg this should all be tightened up at this point, PTAL

File: dashboard/test/models/user_test.rb [hunk 8]
```diff
@@ -1046,6 +1046,61 @@ def complete_script_for_user(user, script, completed_date = Time.now)
...
+  test 'teacher_managed_account? is false for teacher' do
+    refute @teacher.teacher_managed_account?
+  end
+
+  test 'teacher_managed_account? is false for normal student account with hashed email and password' do
+    refute @student.teacher_managed_account?
+  end
+
+  test 'teacher_managed_account? is false for student account in section with oauth connection' do
+    student_with_oauth = create(:student, encrypted_password: nil, provider: 'facebook', uid: '1111111')
```
Comment (ashercodeorg) [inline dashboard/test/models/user_test.rb:1058]: TODO(asher): Create an oauth factory as a follow-up to this PR, making use of it here.

File: dashboard/test/models/user_test.rb [hunk 9]
```diff
@@ -1046,6 +1046,61 @@ def complete_script_for_user(user, script, completed_date = Time.now)
...
+
+    # join picture section
+    picture_section = create(:section, login_type: Section::LOGIN_TYPE_PICTURE)
+    create(:follower, student_user: student_with_oauth, section: picture_section)
+    student_with_oauth.reload
+    refute student_with_oauth.teacher_managed_account?
```
Comment (ashercodeorg) [inline dashboard/test/models/user_test.rb:1064]: Is there a reason the section should not be ripped from the test?

File: dashboard/test/models/user_test.rb [hunk 10]
```diff
@@ -1046,6 +1046,61 @@ def complete_script_for_user(user, script, completed_date = Time.now)
...
+  test 'teacher_managed_account? is true for user account with password but no e-mail' do
+    # These types of accounts happen when teachers created username/password accounts
+    # without e-mails for students (this is no longer allowed)
+    student_with_password_no_email = create(
+      :student,
+      encrypted_password: '123456',
+      email: nil,
+      hashed_email: nil,
+      provider: 'manual'
+    )
```
Comment (ashercodeorg) [inline dashboard/test/models/user_test.rb]: Our DB requires `email` to be non-NULL. Presumably, the email should be set to the empty string?

File: dashboard/test/models/user_test.rb [hunk 11]
```diff
@@ -1046,6 +1046,61 @@ def complete_script_for_user(user, script, completed_date = Time.now)
...
+    picture_section = create(:section, login_type: Section::LOGIN_TYPE_PICTURE)
+    word_section = create(:section, login_type: Section::LOGIN_TYPE_WORD)
+
+    [picture_section, word_section].each do |section|
+      student_with_password = create(:student, encrypted_password: '123456')
+      create(:follower, student_user: student_with_password, section: section)
+      student_with_password.reload
+      refute student_with_password.teacher_managed_account?
+    end
+  end
```
Comment (ashercodeorg) [inline dashboard/test/models/user_test.rb]: Is this test still relevant, given that `teacher_managed_account?` does not care about sections? Also below.

File: dashboard/test/models/user_test.rb (hunk 9 above)
Comment (ewjordan) [reply dashboard/test/models/user_test.rb:1064]: Yes, this is deliberate because I want to make sure that if a student *is* in a section, then even if they don't have a password, it's still not considered teacher-managed if there's an oauth connection. The condition I'd initially used in this PR would have missed this case.

File: dashboard/test/models/user_test.rb (hunk 11 above)
Comment (ewjordan) [reply dashboard/test/models/user_test.rb]: My primary inclination to leave these in is that in other cases, we have historically been looking at section status to figure out things like this (basically, checking if all the sections someone is in are picture or word). Because it took a decent amount of back and forth discussion to realize why that approach doesn't work, I'd tend to leave these in just to make sure we don't regress if someone's refactoring that code further down the line.

Comment (ashercodeorg) [reply dashboard/test/models/user_test.rb]: As you wish. Though I'll maintain that the code and the PR that introduces that code should be sufficient documentation.
